### PR TITLE
Modify the retry flow to go through authentication framework

### PR DESCRIPTION
**Pull request by madurangasiriwardena**
### Proposed changes in this pull request

- Modify the retry flow to go through authentication framework

With this fix,

- Authenticator will retry to authenticate by default.
- If the fail callback function has other steps to execute, authenticator will not retry to authenticate.
- Developers can disable retry for a authentication sequence by setting a parameter in the context.
Ex: `context.retry =  false;`

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/AbstractApplicationAuthenticator.java [hunk 1]
```diff
@@ -62,10 +64,13 @@ public AuthenticatorFlowStatus process(HttpServletRequest request,
         // if an authentication flow
         if (!context.isLogoutRequest()) {
             if (!canHandle(request)
-                || (request.getAttribute(FrameworkConstants.REQ_ATTR_HANDLED) != null && ((Boolean) request
-                    .getAttribute(FrameworkConstants.REQ_ATTR_HANDLED)))) {
+                    || (context.getProperty(FrameworkConstants.RETRYING_STEP) != null
+                        && ((Boolean) context.getProperty(FrameworkConstants.RETRYING_STEP)))
```
Comment (pulasthi7) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/AbstractApplicationAuthenticator.java]: Shall we use "Boolean.TRUE.equals(FrameworkConstants.RETRYING_STEP)"? will avoid the class cast exceptions also.

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/sequence/impl/GraphBasedSequenceHandler.java [hunk 2]
```diff
@@ -144,24 +144,66 @@ private void handleEndOfSequence(HttpServletRequest request, HttpServletResponse
...
+            if (node != null && node.getPrevious() != null) {
+                AuthGraphNode previous = node.getPrevious();
+                if (previous instanceof DynamicDecisionNode && ((DynamicDecisionNode) previous).getPrevious() != null) {
+                    stepNode = ((DynamicDecisionNode) previous).getPrevious();
+                    if (stepNode instanceof StepConfigGraphNode) {
+                        resetGraphForRetrying(request, context, (StepConfigGraphNode)stepNode);
+                        return;
+                    }
+                } else if (previous instanceof StepConfigGraphNode) {
+                    resetGraphForRetrying(request, context, (StepConfigGraphNode)previous);
```
Comment (madawas) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/sequence/impl/GraphBasedSequenceHandler.java]: Formatting issue in casting parameter

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/sequence/impl/GraphBasedSequenceHandler.java [hunk 3]
```diff
@@ -144,24 +144,66 @@ private void handleEndOfSequence(HttpServletRequest request, HttpServletResponse
...
         context.setRequestAuthenticated(false);
         context.getSequenceConfig().setCompleted(true);
         //TODO:Need to consider sending error message to provided error page Uri instead of oauth redirect Uri
     }
 
+    private boolean isRetryEnabled(AuthenticationContext context) {
+
+        if (context.getProperty(FrameworkConstants.JSAttributes.JS_RETRY_STEP) == null) {
+            // If the retry property is not set in the authentication context, it is assumed to be true.
+            return true;
```
Comment (pulasthi7) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/sequence/impl/GraphBasedSequenceHandler.java]: what would happen if the authenticators don't support retrying? Will it force them to retry?

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/sequence/impl/GraphBasedSequenceHandler.java [hunk 4]
```diff
@@ -226,6 +268,24 @@ private boolean handleAuthenticationStep(HttpServletRequest request, HttpServlet
             stepConfig.setSubjectIdentifierStep(false);
         }
 
+        if (flowStatus != null && !(stepConfigGraphNode.getNext() instanceof DynamicDecisionNode)) {
+            switch (flowStatus) {
+                case SUCCESS_COMPLETED:
+                    if (stepConfigGraphNode.getNext() instanceof FailNode) {
+                        FailNode failNode = (FailNode)stepConfigGraphNode.getNext();
+                        JsGraphBuilder.infuse(stepConfigGraphNode,
+                                failNode.getNext() != null ? failNode.getNext():new EndStep());
```
Comment (madawas) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/sequence/impl/GraphBasedSequenceHandler.java]: Formatting issue

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/sequence/impl/GraphBasedSequenceHandler.java [hunk 5]
```diff
@@ -251,9 +311,18 @@ private void handleDecisionPoint(HttpServletRequest request, HttpServletResponse
             switch (flowStatus) {
             case SUCCESS_COMPLETED:
                 executeFunction("success", dynamicDecisionNode, context);
+                if (dynamicDecisionNode.getDefaultEdge() instanceof FailNode) {
+                    FailNode failNode = (FailNode)dynamicDecisionNode.getDefaultEdge();
+                    JsGraphBuilder.infuse(dynamicDecisionNode,
+                            failNode.getNext() != null ? failNode.getNext():new EndStep());
```
Comment (madawas) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/sequence/impl/GraphBasedSequenceHandler.java]: Formatting issue

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultAuthenticationRequestHandler.java [hunk 6]
```diff
@@ -191,7 +191,7 @@ private void handleRememberMeOptionFromLoginPage(HttpServletRequest request, Aut
 
         String rememberMe = request.getParameter(FrameworkConstants.RequestParams.REMEMBER_ME);
 
-        if (rememberMe != null && "on".equalsIgnoreCase(rememberMe)) {
+        if ("on".equalsIgnoreCase(rememberMe)) {
```
Comment (emswbandara) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultAuthenticationRequestHandler.java]: shall we define a constant for this?

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/sequence/impl/GraphBasedSequenceHandler.java (hunk 3 above)
Comment (madurangasiriwardena) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/sequence/impl/GraphBasedSequenceHandler.java]: Created https://github.com/wso2/product-is/issues/3161 to track the issue.

**Pull request merged**